node/core/approval-voting/src/backend.rs

```diff
@@ -0,0 +1,305 @@
+	fn write<I>(&mut self, ops: I) -> SubsystemResult<()>
+		where I: IntoIterator<Item = BackendWriteOp>
+	{
+		let mut tx = DBTransaction::new();
+		for op in ops {
+			match op {
+				BackendWriteOp::WriteStoredBlockRange(stored_block_range) => {
+					tx.put_vec(
+						self.config.col_data,
+						&STORED_BLOCKS_KEY,
+						stored_block_range.encode(),
+					);
+				}
+				BackendWriteOp::WriteBlocksAtHeight(h, blocks) => {
+					tx.put_vec(
+						self.config.col_data,
+						&blocks_at_height_key(h),
+						blocks.encode(),
+					);
+				}
+				BackendWriteOp::DeleteBlocksAtHeight(h) => {
+					tx.delete(
+						self.config.col_data,
+						&blocks_at_height_key(h),
+					);
+				}
+				BackendWriteOp::WriteBlockEntry(block_entry) => {
+					let block_entry: approval_db::v1::BlockEntry = block_entry.into();
+					tx.put_vec(
+						self.config.col_data,
+						&block_entry_key(&block_entry.block_hash),
+						block_entry.encode(),
+					);
+				}
+				BackendWriteOp::DeleteBlockEntry(hash) => {
+					tx.delete(
+						self.config.col_data,
+						&block_entry_key(&hash),
+					);
+				}
+				BackendWriteOp::WriteCandidateEntry(candidate_entry) => {
+					let candidate_entry: approval_db::v1::CandidateEntry = candidate_entry.into();
+					tx.put_vec(
+						self.config.col_data,
+						&candidate_entry_key(&candidate_entry.candidate.hash()),
+						candidate_entry.encode(),
+					);
+				}
+				BackendWriteOp::DeleteCandidateEntry(candidate_hash) => {
+					tx.delete(
+						self.config.col_data,
+						&candidate_entry_key(&candidate_hash),
+					);
+				}
+			}
+		}
+
+		self.inner.write(tx)?;
+		Ok(())
+	}
+}
+
+/// An abstraction over backend storage for the logic of this subsystem.
+pub(super) trait Backend {
+	/// Load a block entry from the DB.
+	fn load_block_entry(&self, hash: &Hash) -> SubsystemResult<Option<BlockEntry>>;
+	/// Load a candidate entry from the DB.
+	fn load_candidate_entry(&self, candidate_hash: &CandidateHash) -> SubsystemResult<Option<CandidateEntry>>;
+	/// Load all blocks at a specific height.
+	fn load_blocks_at_height(&self, height: &BlockNumber) -> SubsystemResult<Vec<Hash>>;
+	/// Load all block from the DB.
+	fn load_all_blocks(&self) -> SubsystemResult<Vec<Hash>>;
+	/// Load stored block range form the DB.
+	fn load_stored_blocks(&self) -> SubsystemResult<Option<StoredBlockRange>>;
+	/// Atomically write the list of operations, with later operations taking precedence over prior.
+	fn write<I>(&mut self, ops: I) -> SubsystemResult<()>
+		where I: IntoIterator<Item = BackendWriteOp>;
+}
+
+/// An in-memory overlay over the backend.
+///
+/// This maintains read-only access to the underlying backend, but can be
+/// converted into a set of write operations which will, when written to
+/// the underlying backend, give the same view as the state of the overlay.
+pub(super) struct OverlayedBackend<'a, B: 'a> {
+	inner: &'a B,
+
+	// `None` means unchanged
+	stored_block_range: Option<StoredBlockRange>,
```

Comment:
Not a fan of this implicit encoding set with `Option`s, could we not store `BackendWriteOp` directly? It costs a few extra bytes for the hash being present in addition, but from a clarity pov this would be advantageous.

Reply:
That's actually a great suggestion... I  would suggest we merge this PR and then follow-up in an additional PR, but more than happy to explore the right approach in this PR.

Paging @rphmeier

Reply:
So basically, instead of a HashMap<K, Option<V>>, you are suggesting we use HashMap<K, BackendWriteOp>?

Reply:
Would the idea there be to have a single map `K -> BackendWriteOp`? Otherwise I think it could get less clear, not more, because the type system would then express the possibility to have a `BackendWriteOp::WriteCandidateEntry` in the `block_entries` map, for example.

Reusing `Option` for this is a little semantically wrong, but maybe a new enum is a better solution:

```rust
enum OverlayValue<T> { Present(T), Deleted }

// Semantically - `None` means 'defer to underlying'
stored_block_range: Option<OverlayValue<StoredBlockRange>>,

// Same here, when calling 'get' for a particular `Hash`.
block_entries: HashMap<Hash, OverlayValue<BlockEntry>>,
```